File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.java
import java.util.*;

class Solution {  
    
    public String solution(String p) {
        return makeCorrect(p);
    }
    private String makeCorrect(String p) {
        if (p.isEmpty()) {
            return "";
        }
        
        String[] split = splitInput(p);
        String u = split[0];
        String v = split[1];
        if (isCorrect(u)) {
            return u + makeCorrect(v);
        } else {
            u = makeCorrectString(u);
           return "(" + makeCorrect(v) + ")" + u;
        }
    }
    
    private boolean isCorrect(String p) {
        char[] chars = p.toCharArray();
        Stack<Character> stack = new Stack<>();
        
        for (char c : chars) {
            if (c == '(') {
                stack.push(c);
                continue;
            } 
            
            if (c == ')') {
                if (stack.isEmpty() || stack.peek() == ')') {
                    return false;
                }
                stack.pop();
            }
        }
        
        if (stack.isEmpty()) {
            return true;
        }
        
        return false;
    }
    
    private String[] splitInput(String p) {
        String[] split = new String[2];
        
        char[] chars = p.toCharArray();
        int left = 0;
        int right = 0;
        int index = 0;
        
        for (char c : chars) {
            if (left != 0 && right != 0 && left == right) {
                break;
            }
            
            if (c == '(') {
                left++;
            }
            if (c == ')') {
                right++;
            }
            index++;
        }
        
        split[0] = p.substring(0, index);
        split[1] = p.substring(index);
        
        return split;
    }
    
    private String makeCorrectString(String u) {
        String removeFirstLast = u.substring(1, u.length() - 1);
        char[] chars = removeFirstLast.toCharArray();
        StringBuilder sb = new StringBuilder();
        for (char c : chars) {
            if (c == '(') {
                sb.append(")");
            } else {
                sb.append("(");
            }
        }
        
        return sb.toString();
    }
}

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.py
def solution(p):
    return make_correct(p)

def make_correct(string):
    if len(string) == 0:
        return ""
    
    u, v = split_u_and_v(string)
    print("u ", u, " v ", v)

    if is_correct(u):
        change = make_correct(v)
        print("u + change ", u + change)
        return u + change
    else:
        new_string = "("
        new_string += make_correct(v)
        new_string += ")"

        split_u = u[1:len(u) - 1]
        print("split_u ", split_u)
        print(new_string)
        for s in split_u:
            if s == "(":
                new_string += ")"
            elif s == ")":
                new_string += "("
        print("correct_u ", new_string)
        return new_string

def split_u_and_v (string):
    left_count = 0
    right_count = 0
    index = 0

    for i in range(len(string)):
        if string[i] == "(":
            left_count += 1
        elif string[i] == ")":
            right_count += 1

        if left_count == right_count:
            index = i
            break
            
    return string[0 : index + 1], string[index + 1 :]

def is_correct(string):
    stack = []

    for s in string:
        if s == "(":
            stack.append(s)
        else:
            if len(stack) == 0:
                return False
            pre = stack.pop()

            if pre == ")":
                return False

    return len(stack) == 0